tests/integration_tests/functional/test_drive_vhost_user.py

```diff
@@ -62,3 +121,270 @@ def test_vhost_user_block(microvm_factory, guest_kernel, rootfs_ubuntu_22):
+    vm.add_vhost_user_drive("rootfs", VHOST_USER_SOCKET, is_root_device=True)
+    vm.add_net_iface()
+    vm.start()
+
+    # Attempt to connect to the VM.
+    # Verify if guest can run commands.
+    exit_code, _, _ = vm.ssh.run("ls")
+    assert exit_code == 0
+
+    # Killing the backend
+    _backend.kill()
+
+    # Verify that Firecracker is still responsive
+    _config = vm.api.vm_config.get().json()
+
+
+def test_device_ordering(microvm_factory, guest_kernel, rootfs_ubuntu_22):
+    """
+    Verify device ordering.
+
+    The root device should correspond to /dev/vda in the guest and
+    the order of the other devices should match their configuration order.
+    """
+
+    VHOST_USER_SOCKET_1 = "/vub_1.socket"
+    VHOST_USER_SOCKET_2 = "/vub_2.socket"
+
+    vm = microvm_factory.build(guest_kernel, None, monitor_memory=False)
+
+    # Converting path from tmpfs ("./srv/..") to local
+    # path on the host ("../build/..")
+    rootfs_path = utils.to_local_dir_path(str(rootfs_ubuntu_22))
+    # Launching vhost-user-block backend
+    _backend = spawn_vhost_user_backend(vm, rootfs_path, VHOST_USER_SOCKET_1, True)
+
+    # We need to setup ssh keys manually because we did not specify rootfs
+    # in microvm_factory.build method
+    ssh_key = rootfs_ubuntu_22.with_suffix(".id_rsa")
+    vm.ssh_key = ssh_key
+    vm.spawn()
+    vm.basic_config(add_root_device=False)
+    vm.add_net_iface()
+
+    # Adding first block device.
+    fs1 = drive_tools.FilesystemFile(os.path.join(vm.fsfiles, "scratch1"), size=128)
+    vm.add_drive("scratch1", fs1.path)
+
+    # Adding second block device (rootfs)
+    vm.add_vhost_user_drive("rootfs", VHOST_USER_SOCKET_1, is_root_device=True)
+
+    # Adding third block device.
+    fs2 = drive_tools.FilesystemFile(os.path.join(vm.fsfiles, "scratch2"), size=512)
+    vm.add_drive("scratch2", fs2.path)
+
+    # Launching vhost-user-block backend
+    _backend2 = spawn_vhost_user_backend(vm, rootfs_path, VHOST_USER_SOCKET_2, True)
+    # Adding forth block device.
+    vm.add_vhost_user_drive("dummy_rootfs", VHOST_USER_SOCKET_2)
+
+    vm.start()
+
+    # Determine the size of the microVM rootfs in bytes.
+    rc, stdout, stderr = utils.run_cmd(
+        "du --apparent-size --block-size=1 {}".format(rootfs_ubuntu_22),
+    )
+    assert rc == 0, f"Failed to get microVM rootfs size: {stderr}"
+
+    assert len(stdout.split()) == 2
+    rootfs_size = stdout.split("\t")[0]
+
+    # The devices were added in this order: fs1, rootfs, fs2.
+    # However, the rootfs is the root device and goes first,
+    # so we expect to see this order: rootfs, fs1, fs2.
+    # The devices are identified by their size.
+    ssh_connection = vm.ssh
+    _check_block_size(ssh_connection, "/dev/vda", rootfs_size)
+    _check_block_size(ssh_connection, "/dev/vdb", fs1.size())
+    _check_block_size(ssh_connection, "/dev/vdc", fs2.size())
+    _check_block_size(ssh_connection, "/dev/vdd", rootfs_size)
```

Comment:
checking both against `rootfs_size` isn't completely robust. What if they're swapped?

Reply:
made last vhost-user drive `rw`, and added an additional testing based on the `ro`/`rw` status.